pgcrypto/fields.py

```diff
@@ -44,77 +33,103 @@ class TextHMACField(HashMixin, models.TextField):
     """DateTime PGP public key encrypted field for postgres."""
+    encrypt_sql = PGP_PUB_ENCRYPT_SQL_WITH_NULLIF
+    cast_type = 'TIMESTAMP'
 
 
-DateTimePGPPublicKeyField.register_lookup(DateTimeExactLookup)
-DateTimePGPPublicKeyField.register_lookup(DateTimeLtLookup)
-DateTimePGPPublicKeyField.register_lookup(DateTimeLteLookup)
-DateTimePGPPublicKeyField.register_lookup(DateTimeGtLookup)
-DateTimePGPPublicKeyField.register_lookup(DateTimeGteLookup)
-DateTimePGPPublicKeyField.register_lookup(DateTimeRangeLookup)
-
-
-class EmailPGPSymmetricKeyField(EmailPGPSymmetricKeyFieldMixin, models.EmailField):
+class EmailPGPSymmetricKeyField(RemoveMaxLengthValidatorMixin,
+                                PGPSymmetricKeyFieldMixin, models.EmailField):
     """Email PGP symmetric key encrypted field."""
 
 
 class IntegerPGPSymmetricKeyField(PGPSymmetricKeyFieldMixin, models.IntegerField):
     """Integer PGP symmetric key encrypted field."""
-    encrypt_sql = INTEGER_PGP_SYM_ENCRYPT_SQL
+    encrypt_sql = PGP_SYM_ENCRYPT_SQL_WITH_NULLIF
     cast_type = 'INT4'
 
 
 class TextPGPSymmetricKeyField(PGPSymmetricKeyFieldMixin, models.TextField):
     """Text PGP symmetric key encrypted field for postgres."""
 
 
-class DatePGPSymmetricKeyField(DatePGPSymmetricKeyFieldMixin, models.TextField):
+class DatePGPSymmetricKeyField(PGPSymmetricKeyFieldMixin, models.DateField):
     """Date PGP symmetric key encrypted field for postgres."""
+    encrypt_sql = PGP_SYM_ENCRYPT_SQL_WITH_NULLIF
+    cast_type = 'DATE'
 
 
-DatePGPSymmetricKeyField.register_lookup(DateTimeExactLookup)
-DatePGPSymmetricKeyField.register_lookup(DateTimeLtLookup)
-DatePGPSymmetricKeyField.register_lookup(DateTimeLteLookup)
-DatePGPSymmetricKeyField.register_lookup(DateTimeGtLookup)
-DatePGPSymmetricKeyField.register_lookup(DateTimeGteLookup)
-DatePGPSymmetricKeyField.register_lookup(DateTimeRangeLookup)
+class DateTimePGPSymmetricKeyField(PGPSymmetricKeyFieldMixin, models.DateTimeField):
+    """DateTime PGP symmetric key encrypted field for postgres."""
+    encrypt_sql = PGP_SYM_ENCRYPT_SQL_WITH_NULLIF
+    cast_type = 'TIMESTAMP'
 
 
-class DateTimePGPSymmetricKeyField(DateTimePGPSymmetricKeyFieldMixin, models.TextField):
-    """DateTime PGP symmetric key encrypted field for postgres."""
+class DecimalPGPPublicKeyField(PGPPublicKeyFieldMixin, models.DecimalField):
+    """Decimal PGP public key encrypted field for postgres."""
+    cast_type = 'NUMERIC(%(max_digits)s, %(decimal_places)s)'
+
+    def get_cast_sql(self):
+        """Get cast sql."""
+        return self.cast_type % {
+            'max_digits': self.max_digits,
+            'decimal_places': self.decimal_places
+        }
+
+
+class DecimalPGPSymmetricKeyField(PGPSymmetricKeyFieldMixin, models.DecimalField):
+    """Decimal PGP symmetric key encrypted field for postgres."""
+    cast_type = 'NUMERIC(%(max_digits)s, %(decimal_places)s)'
+
+    def get_cast_sql(self):
```

Comment:
`get_cast_sql` appears to be duplicated between `DecimalPGPPublicKeyField` and `DecimalPGPSymmetricKeyField`. Consider creating a `DecimalPGPFieldMixin` to remove the duplication.

Reply:
Done. I was trying to not have mixin proliferation which we were having before, but I agree and created a `DecimalPGPFieldMixin`.